Adds script to auto create libvirt podvm config - #435

Saripalli-lavanya opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**
Added script to Automatically create libvirt pool and volume on KVM host and upon successful creation same script will create peer-pod secret, config map & ssh secret. - #435
**- What I did**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

savitrilh commented:
@Saripalli-lavanya Add the script to cleanup when kataconfig is unconfigured

gkurz commented:
Hi @Saripalli-lavanya and @savitrilh ! Thanks for all the huge work you are doing :smile:

I'll just suggest that you break down the first commit into smaller logical and bisectable pieces. Even if this might look like extra work, it is largely beneficial for everyone :
- building a patch series incrementally allows contributors to spot bugs very early, often before even creating the PR
- it is way easier for reviewers to provide valuable feedback and speed up the acceptance of the PR
- subtle bugs can be found by bisecting, which is way more efficient than code inspection

Also, we don't have pre-merge CI currently : bugs will be found late. Being able to bisect can really make a difference for non-obvious ones.

Cheers !

Saripalli-lavanya committed:
Include script to setup libvirt

Added script to that will help create and clean up libvirt configurations needed for Sandboxed containers operator.

Signed-off-by: Saripalli Lavanya <Saripalli.Lavanya@ibm.com>

and

Signed-off-by: savitrilh <savitrilh@in.ibm.com>

Saripalli-lavanya committed:
Added script to setup peer-pod config

Sandboxed containers operator to setup peer-pod config

Signed-off-by: Saripalli Lavanya <Saripalli.Lavanya@ibm.com>

and

Signed-off-by: savitrilh <savitrilh@in.ibm.com>

Saripalli-lavanya committed:
Added job to cleanup peer-pod config setup

Sandboxed containers operator to cleanup peer-pod config

Signed-off-by: Saripalli Lavanya <Saripalli.Lavanya@ibm.com>

and

Signed-off-by: savitrilh <savitrilh@in.ibm.com>

Saripalli-lavanya committed:
Add libvirt dependency installation and pre-configuration functions in setup script

Updated builder script to include installation of dependencies and libvirt setup script which helps configure libvirt and cleanup.

Signed-off-by: Saripalli Lavanya <Saripalli.Lavanya@ibm.com>

and

Signed-off-by: savitrilh <savitrilh@in.ibm.com>

Saripalli-lavanya committed:
Implement methods for getting libvirt-related secrets and config map

Added functions to retrieve ssh-key-secret, ocp-libvirt-secret, and libvirt-podvm-image config map in the operator.

Signed-off-by: Saripalli Lavanya <Saripalli.Lavanya@ibm.com>

and

Signed-off-by: savitrilh <savitrilh@in.ibm.com>

Saripalli-lavanya committed:
Add methods for libvirt configuration and post cleanup job management

Added constants for job statuses and implemented functions for pre-configuration and post-unconfiguration tasks in the libvirt setup.

Saripalli-lavanya committed:
Enhance Libvirt configuration handling and cleanup

Refactored configuration checks and cleanup handling in the confidential handler and OpenShift controller.

bpradipt commented on controllers/image_generator.go:
```diff
@@ -59,6 +59,8 @@ const (
 	LibvirtProvider               = "libvirt"
 	peerpodsImageJobsPathLocation = "/config/peerpods/podvm"
 	azureImageGalleryPrefix       = "PodVMGallery"
+	preConfigJobName              = "osc-podvm-pre-configuration"
```
On a generic note, I'm really not in favour of adding more jobs to the already complicated image generation code. 
Can't the jobs be run in a standalone fashion  (admin runs oc apply -f <job.yaml> or a script) to start with without code integration ?

savitrilh replied:
@bpradipt  If we provide a builder image built with these scripts in the job, the OCP admin can run this job as a standalone process. We believe that the creation and deletion of ConfigMaps, Secrets, and Libvirt pools/volumes can be automated with OSC. We attempted to synchronize this with the Image Job and the code. Additionally, we enhanced it to be scalable so that other providers can also manage their ConfigMaps and Secrets by enhancing provider specific data.

bpradipt replied:
Can't the configmap and secret updates be de-coupled from image generation ?

savitrilh replied:
@bpradipt  Certainly possible to de-couple configmap and secret from image generation
This can be implemented as below:

File config/peerpods/podvm/libvirt-podvm-secret.yaml contains all user Input.
Add new Key/Flag in the config/peerpods/podvm/libvirt-podvm-secret.yaml file as below:
`IMAGE_GEN ="YES"`
This is an option parameter, by default it can be considered as `Yes` . If Admin user doesn't wish to generate image, then it can be set to "NO". Then it will not create image config map and pool, vol.
By default, it can create peer-pods-cm configmap and secrets.

If this option is fine, we can implement and drop that code changes

bpradipt replied:
Let's hear what other maintainers have to say before you make any changes. 

The related issue is quite old and lots have changed since then. So I think some discussion is needed on the linked issue - https://github.com/openshift/sandboxed-containers-operator/issues/435

savitrilh replied:
Thanks @bpradipt  

Please let me know where I need to initiate this discussion. I will initiate the discussion

littlejawa commented on config/peerpods/podvm/libvirt-config-manager.sh:
```diff
@@ -0,0 +1,380 @@
+
+        if [ "\$POOL_EXISTS" -eq 0 ] && [ "\$VOL_EXISTS" -eq 0 ]; then
+            echo "A Libvirt pool named '${libvirt_pool}' with volume '${libvirt_vol_name}' already exists on the KVM host. Please choose a different name."
+            exit 0
```
There is no way to re-use an existing pool?
No big deal if it's a mandatory requirement, but if we can keep a little bit of flexibility, that would be nice - if only for development/test setups :-)
